GPII-1241: Initial implementation of pouchdb-lucene test harness...

the-t-in-rtf opened the pull request:
Per the recent discussions in [GPII-1241](https://issues.gpii.net/browse/GPII-1241), I have spiked a test harness for our lucene search integration.  This pull request covers the initial work.

amb26 commented on tests/js/lucene-tests.js:
```diff
@@ -0,0 +1,138 @@
+
+fluid.defaults("gpii.pouch.lucene.tests.caseHolder", {
+    gradeNames: ["fluid.test.testCaseHolder", "autoInit"],
+    // TODO: Review with Antranig.  We are use `assertLeftHand` for these, but the left-handedness is too shallow to inspect actual records.
```
looks like this comment is now outdated

the-t-in-rtf replied:
Fixed this, and the comment about cleaning up after pouch, which has been fixed after the latest update to `gpii-pouch`.

amb26 commented on package.json:
```diff
@@ -0,0 +1,38 @@
+  "devDependencies": {
+    "adm-zip":              "^0.4.7",
+    "gpii-express":         "git://github.com/GPII/gpii-express#d2b57d7920270926a0fad77c81a39541a2a1a6e6",
+    "gpii-pouch":           "git://github.com/the-t-in-rtf/gpii-pouchdb.git#55b8b3bafb170183985cc529cf84bf00caee5fae",
```
git returns "bad object" when trying to reference this - perhaps it needs to be updated?

the-t-in-rtf replied:
Done.  In doing so I discovered GPII-1259, the fixes for which are now in flight.

amb26 commented on package.json:
```diff
@@ -0,0 +1,38 @@
+    "postinstall": "grunt dedupe-infusion && npm run clone && npm run checkout && npm run build"
+  },
+  "author": "Tony Atkins <tony@raisingthefloor.org>",
+  "license": "See LICENSE file.",
```
Modern versions of npm require this to be a valid "SPDX licence expression"

the-t-in-rtf replied:
Done.

amb26 commented on package.json:
```diff
@@ -0,0 +1,38 @@
+{
```
Modern npm warns about the absence of a "repository" field

the-t-in-rtf replied:
Done.

amb26 commented on src/js/gpii-pouchdb-lucene.js:
```diff
@@ -0,0 +1,140 @@
+            args:     ["{that}"]
+        },
+        // TODO:  Review with Antranig and find a way to handle this bit of shutdown cleanly.
+        "onDestroy.stopProcess": {
```
This should work without problems - I can't run the npm installer on Windows because of the following failure:

> rm -rf $npm_package_config_srcDir && git clone $npm_package_config_url $npm_package_config_srcDir

fatal: repository '$npm_package_config_url' does not exist

Can you tell me what you observe during this stopProcess execution?

the-t-in-rtf replied:
We can pair on the actual error, but I did some research on the install fail and wanted to comment.

Basically, I can't find a safe way to call unix-like commands or even refer to config variables from within `npm run`.  The install will always fail on Windows as a result.

The solution would seem to be to migrate the checkout and maven build to Grunt (there are plugins for both).

the-t-in-rtf commented:
Just to confirm the current status, this needs to be updated to perform its build tasks correctly in Windows.

the-t-in-rtf commented:
Just to summarize, the feedback request has been addressed, but there are still intermittent problems on Windows that Antranig and I will review together.

amb26 commented:
Running tests tests hangs for me:

2016-01-25 16:30:21,058 INFO [Config] Index output goes to: C:\Users\Bosmon\AppData\Local\Temp\75fedxt7-146\couchdb-lucene-1.0.2\indexes
2016-01-25 16:30:21,101 INFO [Main] Accepting connections with SelectChannelConnector@localhost:3598
16:30:21.351:  pouchdb-lucene started successfully...
16:30:21.353:  Sending a GET request to: /local/sample/_design/lucene/by_content?q=cat on port 3598
2016-01-25 16:30:22,514 INFO [sample] Indexing from update_seq start
16:30:22.806:  pouchdb-lucene ready for shutdown...
2016-01-25 16:30:37,586 INFO [sample] View[name=_design/lucene/by_content, digest=bri547edukcwbntm5holofx1e] now at update_seq 5
16:30:42.806:  Test case listener has not responded after 20000ms - at sequence pos 6 of 6 sequence element {
    "listener": "fluid.identity",
    "event": "{testEnvironment}.harness.lucene.events.onShutdownComplete"
} of fixture Testing lucene search integration...

the-t-in-rtf commented:
After long discussion and much testing, I'm happy to confirm that all tests run consistently on Windows.

amb26 commented on src/js/gpii-pouchdb-lucene.js:
```diff
@@ -0,0 +1,253 @@
+    if (that.process) {
+        // Hello, cruel Windows.
+        if (os.platform().indexOf("win") === 0) {
+            var pidLookupCommand = "wmic process where \"name='java.exe' and commandline like '%" + that.id + "%'\" get processid";
```
Please package this as a reusable function accepting the variable material (process name and commandline tag) so that it's ready to be used more generally.

the-t-in-rtf replied:
Done.

amb26 commented on src/js/gpii-pouchdb-lucene.js:
```diff
@@ -0,0 +1,253 @@
+    return "java -Xmx1g -Did=\""+ that.id + "\" -cp " + classpathString + " com.github.rnewson.couchdb.lucene.Main";
+};
+
+//gpii.pouch.lucene.getJavaPath = function () {
```
Leftover funcion

the-t-in-rtf replied:
Done.

amb26 commented on src/js/gpii-pouchdb-lucene.js:
```diff
@@ -0,0 +1,253 @@
+// package.
+//
+"use strict";
+var fluid = fluid || require("infusion");
```
Client-style include

the-t-in-rtf replied:
Done.

amb26 commented on tests/js/launch-test.harness.js:
```diff
@@ -0,0 +1,13 @@
+// A convenience script to start up a copy of the test harness for manual QA.
+var fluid = fluid || require("infusion");
```
Client-style include, etc.

the-t-in-rtf replied:
Done.

amb26 commented on tests/lib/saneresponse.js:
```diff
@@ -0,0 +1,21 @@
+require("../lib/superset");
+
+fluid.registerNamespace("gpii.pouch.lucene.tests");
+gpii.pouch.lucene.tests.isSaneResponse = function (response, body, statusCode, expected) {
```
Presumably this can be eliminated in favour of a gpii-express or kettle function

the-t-in-rtf replied:
There is still some unique wiring, for the common checks I deferred to the comparable function in {{gpii-express}}.

amb26 commented on tests/lib/superset.js:
```diff
@@ -0,0 +1,37 @@
+//
+// Useful in comparing pristine expected output to output that includes automatically generated cruft (IDs, etc.)
+//
+// We use this general pattern in a few places, this one is tailored to the type of results returned by couchdb-lucene.
```
Could you add more details here to illustrate the exact behaviour of the function (perhaps with some examples) to show how the behaviour goes beyond that supported by jqUnit.assertLeftHand

the-t-in-rtf replied:
It is specific to the double-comparison (top-level metadata and then row data) that we need here.  I updated the comments to expand on that.

amb26 commented on .jshintrc:
```diff
@@ -0,0 +1,32 @@
+
+    "node": true,
+
+    "globals": {
```
Remove this block

the-t-in-rtf replied:
Done.

the-t-in-rtf commented:
This one appears to be ready for further review, all outstanding feedback has been addressed.

amb26 commented:
This module's tests fail on a clean install. It depends on a version of gpii-pouch which itself depends on a version of Infusion which is old enough to require deduping. And it no longer runs the dedupe task itself.

13:33:14.552:  ASSERTION FAILED: Error loading infusion - infusion has already been loaded from the path
        E:\source\gits\gpii-pouchdb-lucene\node_modules\infusion
 - please delete the duplicate copy which is found at
        E:\source\gits\gpii-pouchdb-lucene\node_modules\gpii-pouch\node_modules\infusion\src\module

Manually deleting the infusion then runs into a further failure when trying to start up couchdb-lucene:

java.lang.UnsupportedClassVersionError: com/github/rnewson/couchdb/lucene/Main : Unsupported major.minor version 51.0
        at java.lang.ClassLoader.defineClass1(Native Method)

I assume there is a minimum JDK version requirement which should be documented in the README.

the-t-in-rtf commented:
I have addressed the build problems.  You will likely need to remove your `node_modules` directory and run `npm cache clear`, since I also fixed a typo in the package name (it should have been `gpii-pouchdb` instead of `gpii-pouch`).

The README file already mentions that Java 1.7 is required (see line 11).
